mdtk/downloaders.py

```diff
@@ -0,0 +1,233 @@
+"""
+Classes to download data from different source. Each class gets data from one
+source. The base data to get is midi. To contribute a new dataset, create a
+new class which extends DataDownloader, and write an accompanying test in
+./tests/test_downloads.py
+"""
+import os
+import sys
+import shutil
+import urllib
+import warnings
+import zipfile
+import glob
+from tqdm import tqdm
+
+
+
+USER_HOME = os.path.expanduser('~')
+DEFAULT_CACHE_PATH = os.path.join(USER_HOME, '.mdtk_cache')
+
+
+
+def download_file(source, dest, verbose=True, overwrite=None):
+        """Get a file from a url and save it locally"""
+        if verbose:
+            print(f'Downloading {source} to {dest}')
+        if os.path.exists(dest):
+            if overwrite is None:
+                warnings.warn(f'WARNING: {dest} already exists, not '
+                              'downloading', category=UserWarning)
+                return
+            if not overwrite:
+                raise OSError(f'{dest} already exists')
+        try:
+            urllib.request.urlretrieve(source, dest)
+        except urllib.error.HTTPError as e:
+            print(f'Url {source} does not exist', file=sys.stderr)
+            raise e
+
+
+def make_directory(path, overwrite=None, make_parents=True, verbose=True):
+        """Convenience function to create a directory and handle cases where
+        it already exists.
+        
+        Args
+        ----
+        path: str
+            The path of the directory to create
+        overwrite: boolean or None
+            If the path already exists, if overwrite is: True - delete the 
+            existing path; False - return error; None - leave the existing
+            path as it is and throw a warning    
+        make_parents: bool
+            Whether to create the parent folders of a dir if they don't exist
+        verbose: bool
+            Verbosity of printing
+        """
+        if verbose:
+            print(f'Making directory at {path}')
+        
+        if make_parents:
+            mkdir = os.makedirs
+        else:
+            mkdir = os.mkdir
+        
+        try:
+            mkdir(path)
+        except FileExistsError as e:
+            if overwrite is True:
+                print(f'Deleting existing directory: {path}')
+                shutil.rmtree(path)
```

Comment:
(If it is kept), make it explicitly clear somewhere that overwrite=True will delete the directory and all sub-directories within it. It makes sense, but could too easily be overlooked by a potential user.